File: map.js
// Initialize state for all subcategories
let subcategoryState = {};
let activeLegends = {};
let activeTab = 'legend'; // Default tab is legend

let climateData = null; // Will store the seasonal climate data
let currentRegion = null; // Store the currently selected region
let chart = null; // Store the chart instance for reuse/updates

// Load climate data on page load
document.addEventListener('DOMContentLoaded', async function() {
    try {
        const response = await fetch('seasonal_data_2005_2014.json');
        climateData = await response.json();
        console.log('Climate data loaded:', climateData);
        
        // Set up event listeners for analyze buttons in popups
        setupAnalyzeButtons();
        
    } catch (error) {
        console.error('Error loading climate data:', error);
    }
});


// Function to toggle category (expand/collapse)
function toggleCategory(category) {
    // Get the selected category's subcategories
    const targetSubcategories = document.getElementById(`${category}-subcategories`);
    
    // Check if it's already visible
    const isVisible = targetSubcategories && 
                     targetSubcategories.style.display === 'block';
    
    // Get all subcategories elements
    const allSubcategories = document.querySelectorAll('.subcategories');
    
    // Hide all subcategories first
    allSubcategories.forEach(el => {
        el.style.display = 'none';
    });
    
    // If it wasn't visible, show it (otherwise it remains hidden)
    if (targetSubcategories && !isVisible) {
        targetSubcategories.style.display = 'block';
    }
    
    // Highlight the active category button - only if we're opening
    const allCategoryBtns = document.querySelectorAll('.category-btn');
    allCategoryBtns.forEach(btn => {
        btn.classList.remove('active');
    });
    
    // Add active class to the clicked button - only if we're opening
    if (!isVisible) {
        document.querySelector(`.category-btn[onclick="toggleCategory('${category}')"]`).classList.add('active');
    }
}

// Function to toggle a specific subcategory layer
function toggleSubcategoryLayer(checkbox) {
    const layerName = checkbox.dataset.layer;
    const legendType = checkbox.dataset.legend;
    const isChecked = checkbox.checked;
    
    // Store the state
    subcategoryState[checkbox.id] = isChecked;
    
    // Find the corresponding layer control checkbox in Folium's layer control
    const layerControls = document.querySelectorAll('.leaflet-control-layers-overlays input');
    
    layerControls.forEach(input => {
        const label = input.nextElementSibling.textContent.trim();
        if (label === layerName && input.checked !== isChecked) {
            // Programmatically click the Folium layer checkbox
            input.click();
        }
    });
    
    // Update legend visibility
    updateLegendVisibility(legendType, isChecked);
}

// Function to update legend visibility
function updateLegendVisibility(legendType, isChecked) {
    if (!legendType) return;
    
    // Track if this legend should be shown
    if (isChecked) {
        activeLegends[legendType] = true;
    } else {
        // Only set to false if no other layer of this type is active
        const otherLayersOfSameType = document.querySelectorAll(`[data-legend="${legendType}"]`);
        let anyOtherActive = false;
        otherLayersOfSameType.forEach(layer => {
            if (layer.checked && layer !== document.activeElement) {
                anyOtherActive = true;
            }
        });
        
        if (!anyOtherActive) {
            activeLegends[legendType] = false;
        }
    }
    
    // Update legend visibility if we're in the legend tab
    if (activeTab === 'legend') {
        for (const type in activeLegends) {
            const legendElement = document.getElementById(`legend-${type}`);
            if (legendElement) {
                legendElement.style.display = activeLegends[type] ? 'block' : 'none';
            }
        }
    }
}

// Function to switch between tabs (legend and analysis)
function switchTab(tabName) {
    activeTab = tabName;
    
    // Update tab button styles
    document.getElementById('legend-tab').classList.toggle('active', tabName === 'legend');
    document.getElementById('analysis-tab').classList.toggle('active', tabName === 'analysis');
    
    // Show/hide appropriate container
    const legendContainer = document.getElementById('legend-container');
    const analysisContainer = document.getElementById('analysis-container');
    
    if (tabName === 'legend') {
        legendContainer.style.display = 'block';
        analysisContainer.style.display = 'none';
        
        // Update legend items visibility based on active layers
        for (const type in activeLegends) {
            const legendElement = document.getElementById(`legend-${type}`);
            if (legendElement) {
                legendElement.style.display = activeLegends[type] ? 'block' : 'none';
            }
        }
    } else { // analysis
        legendContainer.style.display = 'none';
        analysisContainer.style.display = 'block';
    }
}

// Set up event listeners for the Analyze buttons in map popups
function setupAnalyzeButtons() {
    // Use MutationObserver to detect when popups are added to the DOM
    const observer = new MutationObserver(mutations => {
        mutations.forEach(mutation => {
            mutation.addedNodes.forEach(node => {
                if (node.nodeType === Node.ELEMENT_NODE) {
                    const analyzeButton = node.querySelector('#analyze-button');
                    if (analyzeButton) {
                        analyzeButton.addEventListener('click', function() {
                            // Get the region name from the popup content
                            const popupContent = analyzeButton.closest('.leaflet-popup-content');
                            const regionName = popupContent.querySelector('h3').textContent;
                            
                            // Switch to analysis tab and show data for the region
                            switchTab('analysis');
                            showRegionAnalysis(regionName);
                        });
                    }
                }
            });
        });
    });
    
    // Start observing the document body
    observer.observe(document.body, { childList: true, subtree: true });
}

// Show analysis for the selected region
function showRegionAnalysis(regionName) {
    // Set current region
    currentRegion = regionName;
    
    // Update UI
    document.getElementById('analysis-placeholder').style.display = 'none';
    document.getElementById('analysis-region-data').style.display = 'block';
    document.getElementById('region-title').textContent = `${regionName} Region Analysis`;
    
    // Show FMAM season data by default
    showSeasonData('fmam');
}

// Show seasonal data
function showSeasonData(season) {
    if (!climateData || !currentRegion || !climateData[currentRegion]) {
        console.error('Climate data not available for region:', currentRegion);
        return;
    }
    
    // Highlight the active season button
    document.querySelectorAll('.season-btn').forEach(btn => {
        btn.classList.remove('active');
    });
    document.getElementById(`btn-${season}`).classList.add('active');
    
    // Get data for the selected season
    const seasonData = climateData[currentRegion][season];
    if (!seasonData) {
        console.error(`Data for season ${season} not available`);
        return;
    }
    
    // Prepare the data for the chart
    const years = Array.from({length: seasonData.length}, (_, i) => 2005 + i);
    
    // Create or update the chart
    createBarChart(years, seasonData, `${season.toUpperCase()} Seasonal Rainfall (2005-2014)`, 'Year', 'Rainfall (mm)');
    
    // Display summary statistics
    const average = (seasonData.reduce((a, b) => a + b, 0) / seasonData.length).toFixed(2);
    const min = Math.min(...seasonData).toFixed(2);
    const max = Math.max(...seasonData).toFixed(2);
    
    document.getElementById('data-summary').innerHTML = `
        <h4>Summary Statistics:</h4>
        <p><strong>Average Rainfall:</strong> ${average} mm</p>
        <p><strong>Minimum:</strong> ${min} mm (${years[seasonData.indexOf(Math.min(...seasonData))]})</p>
        <p><strong>Maximum:</strong> ${max} mm (${years[seasonData.indexOf(Math.max(...seasonData))]})</p>
    `;
}

// Show annual cycle data
function showAnnualCycle() {
    if (!climateData || !currentRegion || !climateData[currentRegion]) {
        console.error('Climate data not available for region:', currentRegion);
        return;
    }
    
    // Highlight the active button
    document.querySelectorAll('.season-btn').forEach(btn => {
        btn.classList.remove('active');
    });
    document.getElementById('btn-annual').classList.add('active');
    
    // Prepare data for the annual cycle
    const seasonLabels = ['FMAM', 'JJAS', 'ONDJ'];
    const averagesBySeason = seasonLabels.map(season => {
        const seasonKey = season.toLowerCase();
        const seasonData = climateData[currentRegion][seasonKey];
        return seasonData ? (seasonData.reduce((a, b) => a + b, 0) / seasonData.length) : 0;
    });
    
    // Create or update the chart
    createBarChart(seasonLabels, averagesBySeason, 'Average Rainfall by Season (2005-2014)', 'Season', 'Average Rainfall (mm)');
    
    // Display summary
    const annualAverage = (averagesBySeason.reduce((a, b) => a + b, 0) / averagesBySeason.length).toFixed(2);
    const wetSeason = seasonLabels[averagesBySeason.indexOf(Math.max(...averagesBySeason))];
    const drySeason = seasonLabels[averagesBySeason.indexOf(Math.min(...averagesBySeason))];
    
    document.getElementById('data-summary').innerHTML = `
        <h4>Annual Cycle Summary:</h4>
        <p><strong>Overall Average:</strong> ${annualAverage} mm per season</p>
        <p><strong>Wettest Season:</strong> ${wetSeason} (${Math.max(...averagesBySeason).toFixed(2)} mm)</p>
        <p><strong>Driest Season:</strong> ${drySeason} (${Math.min(...averagesBySeason).toFixed(2)} mm)</p>
    `;
}

// Show time series data
function showTimeSeries() {
    if (!climateData || !currentRegion || !climateData[currentRegion]) {
        console.error('Climate data not available for region:', currentRegion);
        return;
    }
    
    // Highlight the active button
    document.querySelectorAll('.season-btn').forEach(btn => {
        btn.classList.remove('active');
    });
    document.getElementById('btn-timeseries').classList.add('active');
    
    // Prepare data for all seasons
    const years = Array.from({length: 10}, (_, i) => 2005 + i);
    const fmamData = climateData[currentRegion]['fmam'] || [];
    const jjasData = climateData[currentRegion]['jjas'] || [];
    const ondjData = climateData[currentRegion]['ondj'] || [];
    
    // Create or update the time series chart
    createTimeSeriesChart(years, fmamData, jjasData, ondjData);
    
    // Display summary
    const fmamTrend = calculateTrend(fmamData);
    const jjasTrend = calculateTrend(jjasData);
    const ondjTrend = calculateTrend(ondjData);
    
    document.getElementById('data-summary').innerHTML = `
        <h4>Time Series Analysis (2005-2014):</h4>
        <p><strong>FMAM Trend:</strong> ${fmamTrend > 0 ? 'Increasing' : 'Decreasing'} (${Math.abs(fmamTrend).toFixed(2)} mm/year)</p>
        <p><strong>JJAS Trend:</strong> ${jjasTrend > 0 ? 'Increasing' : 'Decreasing'} (${Math.abs(jjasTrend).toFixed(2)} mm/year)</p>
        <p><strong>ONDJ Trend:</strong> ${ondjTrend > 0 ? 'Increasing' : 'Decreasing'} (${Math.abs(ondjTrend).toFixed(2)} mm/year)</p>
    `;
}

// Calculate a simple linear trend
function calculateTrend(data) {
    if (!data || data.length < 2) return 0;
    
    const n = data.length;
    let sumX = 0, sumY = 0, sumXY = 0, sumX2 = 0;
    
    for (let i = 0; i < n; i++) {
        sumX += i;
        sumY += data[i];
        sumXY += i * data[i];
        sumX2 += i * i;
    }
    
    const slope = (n * sumXY - sumX * sumY) / (n * sumX2 - sumX * sumX);
    return slope;
}

// Create or update a bar chart
function createBarChart(labels, data, title, xAxisLabel, yAxisLabel) {
    const ctx = document.getElementById('climate-chart').getContext('2d');
    
    // Destroy previous chart if it exists
    if (chart) {
        chart.destroy();
    }
    
    chart = new Chart(ctx, {
        type: 'bar',
        data: {
            labels: labels,
            datasets: [{
                label: title,
                data: data,
                backgroundColor: 'rgba(54, 162, 235, 0.6)',
                borderColor: 'rgba(54, 162, 235, 1)',
                borderWidth: 1
            }]
        },
        options: {
            responsive: true,
            maintainAspectRatio: false,
            scales: {
                y: {
                    beginAtZero: true,
                    title: {
                        display: true,
                        text: yAxisLabel
                    }
                },
                x: {
                    title: {
                        display: true,
                        text: xAxisLabel
                    }
                }
            },
            plugins: {
                title: {
                    display: true,
                    text: title,
                    font: {
                        size: 16
                    }
                }
            }
        }
    });
}

// Create or update a time series chart
function createTimeSeriesChart(years, fmamData, jjasData, ondjData) {
    const ctx = document.getElementById('climate-chart').getContext('2d');
    
    // Destroy previous chart if it exists
    if (chart) {
        chart.destroy();
    }
    
    chart = new Chart(ctx, {
        type: 'line',
        data: {
            labels: years,
            datasets: [
                {
                    label: 'FMAM Season',
                    data: fmamData,
                    borderColor: 'rgba(54, 162, 235, 1)',
                    backgroundColor: 'rgba(54, 162, 235, 0.1)',
                    borderWidth: 2,
                    tension: 0.1
                },
                {
                    label: 'JJAS Season',
                    data: jjasData,
                    borderColor: 'rgba(75, 192, 192, 1)',
                    backgroundColor: 'rgba(75, 192, 192, 0.1)',
                    borderWidth: 2,
                    tension: 0.1
                },
                {
                    label: 'ONDJ Season',
                    data: ondjData,
                    borderColor: 'rgba(255, 159, 64, 1)',
                    backgroundColor: 'rgba(255, 159, 64, 0.1)',
                    borderWidth: 2,
                    tension: 0.1
                }
            ]
        },
        options: {
            responsive: true,
            maintainAspectRatio: false,
            scales: {
                y: {
                    beginAtZero: true,
                    title: {
                        display: true,
                        text: 'Rainfall (mm)'
                    }
                },
                x: {
                    title: {
                        display: true,
                        text: 'Year'
                    }
                }
            },
            plugins: {
                title: {
                    display: true,
                    text: 'Rainfall Time Series (2005-2014)',
                    font: {
                        size: 16
                    }
                }
            }
        }
    });
}

// Function to sync checkboxes with Folium layer states
function syncLayerStates() {
    const layerControls = document.querySelectorAll('.leaflet-control-layers-overlays input');
    const subcategoryCheckboxes = document.querySelectorAll('.subcategory input[type="checkbox"]');
    
    // Reset active legends
    activeLegends = {
        rainfall: false,
        temperature: false,
        drought: false,
        climate: false
    };
    
    // Create a mapping of layer names to their current state
    const layerStates = {};
    layerControls.forEach(input => {
        const label = input.nextElementSibling.textContent.trim();
        layerStates[label] = input.checked;
    });
    
    // Update subcategory checkboxes based on layer states
    subcategoryCheckboxes.forEach(checkbox => {
        const layerName = checkbox.dataset.layer;
        const legendType = checkbox.dataset.legend;
        
        if (layerName in layerStates) {
            checkbox.checked = layerStates[layerName];
            subcategoryState[checkbox.id] = layerStates[layerName];
            
            // Update legend state
            if (checkbox.checked && legendType) {
                activeLegends[legendType] = true;
            }
        }
    });
    
    // Update all legends visibility if in legend tab
    if (activeTab === 'legend') {
        for (const type in activeLegends) {
            const legendElement = document.getElementById(`legend-${type}`);
            if (legendElement) {
                legendElement.style.display = activeLegends[type] ? 'block' : 'none';
            }
        }
    }
}

// Initialize on page load
document.addEventListener('DOMContentLoaded', function() {
    // Hide all subcategories initially
    const allSubcategories = document.querySelectorAll('.subcategories');
    allSubcategories.forEach(el => {
        el.style.display = 'none';
    });
    
    // Set up tabs
    const legendTabBtn = document.getElementById('legend-tab');
    const analysisTabBtn = document.getElementById('analysis-tab');
    
    if (legendTabBtn) {
        legendTabBtn.addEventListener('click', () => switchTab('legend'));
    }
    
    if (analysisTabBtn) {
        analysisTabBtn.addEventListener('click', () => switchTab('analysis'));
    }
    
    // Initialize tab state
    switchTab('legend');
    
    // Initialize active legends state
    activeLegends = {
        rainfall: false,
        temperature: false,
        drought: false,
        climate: false
    };
    
    // Set up monitoring of Folium layer changes
    setTimeout(() => {
        // Initial sync
        syncLayerStates();
        
        // Add event listeners to Folium layer controls
        const layerControls = document.querySelectorAll('.leaflet-control-layers-overlays input');
        layerControls.forEach(input => {
            input.addEventListener('change', syncLayerStates);
        });
    }, 1000); // Wait for Folium to initialize
    
    // Show tooltip on info icon hover
    const infoIcons = document.querySelectorAll('.info-icon');
    infoIcons.forEach(icon => {
        icon.addEventListener('mouseover', function() {
            const tooltip = document.createElement('div');
            tooltip.className = 'tooltip';
            tooltip.textContent = this.title;
            document.body.appendChild(tooltip);
            
            const rect = this.getBoundingClientRect();
            tooltip.style.top = `${rect.bottom + 5}px`;
            tooltip.style.left = `${rect.left + 5}px`;
            
            this.addEventListener('mouseout', function() {
                document.body.removeChild(tooltip);
            }, { once: true });
        });
    });
});

File: map.py
import folium
import json
from folium.plugins import HeatMap
import branca.colormap as cm

# Initialize the map
mapObj = folium.Map(location=[9.102096738726456, 40.71533203125001], zoom_start=6, zoom_control=False)

# Load the seasonal climate data
with open("seasonal_data_2005_2014.json", "r", encoding="utf-8") as f:
    climate_data = json.load(f)

# Load the Ethiopia GeoJSON file
with open("et.json", "r", encoding="utf-8") as f:
    ethiopia_geojson = json.load(f)

# Ethiopia GeoJSON layer
bordersStyle = {
    'color': 'grey',
    'weight': '2',
    'fillColor': 'purple',
    'fillOpacity': 0.2
}

folium.GeoJson("et.json", name="Ethiopia", style_function=lambda x: bordersStyle).add_to(mapObj)

# Create a feature group for shapes (circles, markers, etc.)
markerLayer = folium.FeatureGroup(name="Marker").add_to(mapObj)

# Improved function to calculate the centroid of a polygon
def calculate_centroid(coordinates):
    """Calculate the centroid of a polygon using the shoelace formula."""
    area = 0
    cx = cy = 0
    
    # For simple polygon with too few points
    if len(coordinates) < 3:
        return sum(p[1] for p in coordinates) / len(coordinates), sum(p[0] for p in coordinates) / len(coordinates)
    
    for i in range(len(coordinates) - 1):
        # Cross product for area calculation
        area_part = coordinates[i][0] * coordinates[i+1][1] - coordinates[i+1][0] * coordinates[i][1]
        area += area_part
        
        # Weighted coordinates
        cx += (coordinates[i][0] + coordinates[i+1][0]) * area_part
        cy += (coordinates[i][1] + coordinates[i+1][1]) * area_part
    
    # Finish the area calculation
    area /= 2
    area = abs(area)  # Make sure area is positive
    
    # Calculate centroid
    if area != 0:
        cx = cx / (6 * area)
        cy = cy / (6 * area)
        return cy, cx  # Return lat, lon
    else:
        # Fallback to average if area is zero
        return sum(p[1] for p in coordinates) / len(coordinates), sum(p[0] for p in coordinates) / len(coordinates)

# Function to calculate bounds-based center (visual center)
def calculate_visual_center(coordinates):
    """Calculate the center of the bounding box of a polygon."""
    # Extract all lon/lat coordinates
    all_lons = [coord[0] for coord in coordinates]
    all_lats = [coord[1] for coord in coordinates]
    
    # Calculate bounds
    min_lon = min(all_lons)
    max_lon = max(all_lons)
    min_lat = min(all_lats)
    max_lat = max(all_lats)
    
    # Return the center of the bounding box
    return (min_lat + max_lat) / 2, (min_lon + max_lon) / 2

# Extract region information from the GeoJSON file and add markers
for feature in ethiopia_geojson['features']:
    # Get region name from properties
    region_name = feature['properties'].get('name', 'Unknown Region')
    
    # Calculate marker placement based on geometry type
    if feature['geometry']['type'] == 'Polygon':
        # For simple polygons
        coordinates = feature['geometry']['coordinates'][0]  # Use outer ring
        # Try mathematical centroid first
        lat, lon = calculate_centroid(coordinates)
        
        # Check if the centroid is within reasonable bounds
        # If not, use visual center as fallback
        min_lon = min(coord[0] for coord in coordinates)
        max_lon = max(coord[0] for coord in coordinates)
        min_lat = min(coord[1] for coord in coordinates)
        max_lat = max(coord[1] for coord in coordinates)
        
        # If centroid is outside the bounding box, use visual center instead
        if not (min_lon <= lon <= max_lon and min_lat <= lat <= max_lat):
            lat, lon = calculate_visual_center(coordinates)
            
    elif feature['geometry']['type'] == 'MultiPolygon':
        # For multipolygons, find the largest polygon and use its centroid
        largest_area = 0
        lat = lon = 0
        
        for polygon in feature['geometry']['coordinates']:
            coords = polygon[0]  # Outer ring of this polygon
            
            # Calculate area to determine the largest polygon
            poly_area = 0
            for i in range(len(coords) - 1):
                poly_area += coords[i][0] * coords[i+1][1] - coords[i+1][0] * coords[i][1]
            poly_area = abs(poly_area / 2)
            
            if poly_area > largest_area:
                largest_area = poly_area
                curr_lat, curr_lon = calculate_centroid(coords)
                
                # Verify the centroid is inside the polygon's bounds
                min_lon = min(coord[0] for coord in coords)
                max_lon = max(coord[0] for coord in coords)
                min_lat = min(coord[1] for coord in coords)
                max_lat = max(coord[1] for coord in coords)
                
                # If outside bounds, use visual center
                if not (min_lon <= curr_lon <= max_lon and min_lat <= curr_lat <= max_lat):
                    curr_lat, curr_lon = calculate_visual_center(coords)
                    
                lat, lon = curr_lat, curr_lon
    else:
        # Skip if not a polygon or multipolygon
        continue
    
    # Add marker at the calculated position
    folium.Marker(
        location=[lat, lon],
        icon=folium.Icon(icon='map-marker', prefix='fa', color='darkblue'),
        tooltip=region_name,
        popup=folium.Popup(f"""<h3>{region_name}</h3> <br/>
                This is the <b>{region_name}</b> region<br/><br/>
                <button id="analyze-button" onclick="analyzeRegion('{region_name}')">Analyze</button> """, max_width=500)
    ).add_to(markerLayer)

# Feature Groups for different hazard types and their subcategories
# Rainfall subcategories
rainfall_weekly_exceptional = folium.FeatureGroup(name="Weekly Exceptional Rainfall", show=False)
rainfall_weekly_total = folium.FeatureGroup(name="Weekly Total Rainfall Forecast", show=False)

# Temperature subcategories
temperature_current = folium.FeatureGroup(name="Current Temperature", show=False)

# Climate Change subcategories
climate_change_trends = folium.FeatureGroup(name="Climate Change Trends", show=False)

# Drought subcategories
drought_current = folium.FeatureGroup(name="Current Drought Conditions", show=False)


# Sample Heatmap Data
total_weekly_rainfall_data = [[7.1, 43.6, 60], [12.0, 40.8, 50], [7.8, 34.4, 70]]
current_temperature_data = [[7.2, 42.5, 35], [11.5, 39.9, 40], [8.0, 33.5, 30]]
current_drought_data = [[7.1, 43.6, 60], [12.0, 40.8, 50], [7.8, 34.4, 70]]

# Load the GeoJSON file for rainfall data
with open("rain.geojson", "r", encoding="utf-8") as f:
    geojson_data = json.load(f)

# Extract coordinates and rainfall values
rainfall_data = []
for feature in geojson_data["features"]:
    try:
        lon, lat = feature["geometry"]["coordinates"]  # Extract coordinates
        rainfall = feature["properties"].get("rainfall", 50)  # Default to 50 if missing
        rainfall_data.append([lat, lon, rainfall])
    except (KeyError, TypeError, ValueError):
        continue  # Skip invalid data

# Add HeatMap layers
HeatMap(data=rainfall_data, radius=20, blur=10).add_to(rainfall_weekly_exceptional)
HeatMap(data=total_weekly_rainfall_data, radius=20, blur=10).add_to(rainfall_weekly_total)
HeatMap(data=current_temperature_data, radius=20, blur=10).add_to(temperature_current)
HeatMap(data=current_drought_data, radius=20, blur=10).add_to(drought_current)

# Add layers to map
mapObj.add_child(rainfall_weekly_exceptional)
mapObj.add_child(rainfall_weekly_total)
mapObj.add_child(temperature_current)
mapObj.add_child(climate_change_trends)
mapObj.add_child(drought_current)

# Add Layer Control
folium.LayerControl().add_to(mapObj)

# Custom sidebar and tab structure (from your original code)
custom_sidebar = """
<link rel="stylesheet" type="text/css" href="map.css">

<!-- Main sidebar with hazard categories -->
<div id="sidebar">
    <h3>Hazard Watch</h3>
        
    <!-- Categories Section -->
    <div class="categories-section">
        <div class="category">
            <button class="category-btn" onclick="toggleCategory('rainfall')">Rainfall</button>
            <div id="rainfall-subcategories" class="subcategories">
                <div class="subcategory">
                    <label class="toggle-switch">
                        <input type="checkbox" id="rainfall-weekly-exceptional" data-layer="Weekly Exceptional Rainfall" data-legend="rainfall" onclick="toggleSubcategoryLayer(this)">
                        <span class="slider"></span>
                    </label>
                    <label for="rainfall-weekly-exceptional">Weekly Exceptional Rainfall</label>
                    <span class="info-icon" title="Latest: 5th Mar 2025 to 12th Mar 2025">ⓘ</span>
                </div>
                <div class="subcategory">
                    <label class="toggle-switch">
                        <input type="checkbox" id="rainfall-weekly-total" data-layer="Weekly Total Rainfall Forecast" data-legend="rainfall" onclick="toggleSubcategoryLayer(this)">
                        <span class="slider"></span>
                    </label>
                    <label for="rainfall-weekly-total">Weekly Total Rainfall Forecast</label>
                    <span class="info-icon" title="Latest: 5th Mar 2025 to 12th Mar 2025">ⓘ</span>
                </div>
            </div>
        </div>
        
        <div class="category">
            <button class="category-btn" onclick="toggleCategory('temperature')">Temperature</button>
            <div id="temperature-subcategories" class="subcategories">
                <div class="subcategory">
                    <label class="toggle-switch">
                        <input type="checkbox" id="temperature-current" data-layer="Current Temperature" data-legend="temperature" onclick="toggleSubcategoryLayer(this)">
                        <span class="slider"></span>
                    </label>
                    <label for="temperature-current">Current Temperature</label>
                    <span class="info-icon" title="Latest available data">ⓘ</span>
                </div>
                <div class="subcategory">
                    <label class="toggle-switch">
                        <input type="checkbox" id="temperature-forecast" data-layer="Temperature Forecast" data-legend="temperature" onclick="toggleSubcategoryLayer(this)">
                        <span class="slider"></span>
                    </label>
                    <label for="temperature-forecast">Temperature Forecast</label>
                    <span class="info-icon" title="7-day forecast">ⓘ</span>
                </div>
            </div>
        </div>
        
        <div class="category">
            <button class="category-btn" onclick="toggleCategory('climateChange')">Climate Change</button>
            <div id="climateChange-subcategories" class="subcategories">
                <div class="subcategory">
                    <label class="toggle-switch">
                        <input type="checkbox" id="climate-change-trends" data-layer="Climate Change Trends" data-legend="climate" onclick="toggleSubcategoryLayer(this)">
                        <span class="slider"></span>
                    </label>
                    <label for="climate-change-trends">Climate Change Trends</label>
                    <span class="info-icon" title="Historical trends">ⓘ</span>
                </div>
                <div class="subcategory">
                    <label class="toggle-switch">
                        <input type="checkbox" id="climate-change-impact" data-layer="Climate Change Impact" data-legend="climate" onclick="toggleSubcategoryLayer(this)">
                        <span class="slider"></span>
                    </label>
                    <label for="climate-change-impact">Climate Change Impact</label>
                    <span class="info-icon" title="Projected impact">ⓘ</span>
                </div>
            </div>
        </div>
        
        <div class="category">
            <button class="category-btn" onclick="toggleCategory('drought')">Drought</button>
            <div id="drought-subcategories" class="subcategories">
                <div class="subcategory">
                    <label class="toggle-switch">
                        <input type="checkbox" id="drought-current" data-layer="Current Drought Conditions" data-legend="drought" onclick="toggleSubcategoryLayer(this)">
                        <span class="slider"></span>
                    </label>
                    <label for="drought-current">Current Drought Conditions</label>
                    <span class="info-icon" title="Current conditions">ⓘ</span>
                </div>
                <div class="subcategory">
                    <label class="toggle-switch">
                        <input type="checkbox" id="drought-forecast" data-layer="Drought Forecast" data-legend="drought" onclick="toggleSubcategoryLayer(this)">
                        <span class="slider"></span>
                    </label>
                    <label for="drought-forecast">Drought Forecast</label>
                    <span class="info-icon" title="3-month forecast">ⓘ</span>
                </div>
            </div>
        </div>
    </div>
</div>

<!-- Tab container for Legend and Analysis -->
<div id="tab-container">
    <!-- Tab Navigation -->
    <div class="tab-navigation">
        <button id="legend-tab" class="tab-btn active" onclick="switchTab('legend')">LEGEND</button>
        <button id="analysis-tab" class="tab-btn" onclick="switchTab('analysis')">ANALYSIS</button>
    </div>

    <!-- Legend Container (Legend Tab) -->
    <div id="legend-container" class="tab-content">
        <div id="legend-rainfall" class="legend-item" style="display: none;">
            <h4>Rainfall (mm)</h4>
            <div class="legend-gradient rainfall-gradient"></div>
            <div class="legend-labels">
                <span>0</span>
                <span>25</span>
                <span>50</span>
                <span>75</span>
                <span>100+</span>
            </div>
        </div>
        <div id="legend-temperature" class="legend-item" style="display: none;">
            <h4>Temperature (°C)</h4>
            <div class="legend-gradient temperature-gradient"></div>
            <div class="legend-labels">
                <span>0</span>
                <span>12.5</span>
                <span>25</span>
                <span>37.5</span>
                <span>50+</span>
            </div>
        </div>
        <div id="legend-drought" class="legend-item" style="display: none;">
            <h4>Drought Severity Index</h4>
            <div class="legend-gradient drought-gradient"></div>
            <div class="legend-labels">
                <span>0</span>
                <span>25</span>
                <span>50</span>
                <span>75</span>
                <span>100</span>
            </div>
        </div>
    </div>

    <!-- Analysis Container (Analysis Tab) -->
    <div id="analysis-container" class="tab-content" style="display: none;">
        <div id="analysis-content">
            <div id="analysis-placeholder">
                <p>Click "Analyze" on a region marker to see climate analysis.</p>
            </div>
            <div id="analysis-region-data" style="display: none;">
                <h3 id="region-title">Region Analysis</h3>
                
                <!-- Seasonal buttons -->
                <div class="analysis-buttons">
                    <button id="btn-fmam" class="season-btn" onclick="showSeasonData('fmam')">FMAM</button>
                    <button id="btn-jjas" class="season-btn" onclick="showSeasonData('jjas')">JJAS</button>
                    <button id="btn-ondj" class="season-btn" onclick="showSeasonData('ondj')">ONDJ</button>
                    <button id="btn-annual" class="season-btn" onclick="showAnnualCycle()">Annual Cycle</button> 
                    <button id="btn-timeseries" class="season-btn" onclick="showTimeSeries()">Time Series</button>
                </div>
                
                <!-- Data visualization container -->
                <div id="data-visualization">
                    <div id="chart-container">
                        <canvas id="climate-chart" width="400" height="300"></canvas>
                    </div>
                    <div id="data-summary"></div>
                </div>
            </div>
        </div>
    </div>
</div>
<script src="https://cdn.jsdelivr.net/npm/chart.js"></script>
<script src="map.js"></script>

<script>
// Add a function to handle region analysis clicks and pass the region name to the map.js functions
function analyzeRegion(regionName) {
    // Update the region title in the analysis tab
    document.getElementById('region-title').textContent = regionName + ' Analysis';
    
    // Hide the placeholder and show the analysis content
    document.getElementById('analysis-placeholder').style.display = 'none';
    document.getElementById('analysis-region-data').style.display = 'block';
    
    // Switch to the analysis tab
    switchTab('analysis');
    
    // You can add more code here to load specific data for the region
    // This would connect to the existing map.js functions
}
</script>
"""

# Add custom HTML to map
mapObj.get_root().html.add_child(folium.Element(custom_sidebar))

# Pass the climate data to JavaScript
folium.Element(f'<script>const seasonalData = {json.dumps(climate_data)};</script>').add_to(mapObj.get_root().html)

# Save map to file
mapObj.save('output.html')